src/harness/test_harness_objects.py

```diff
@@ -0,0 +1,327 @@
+
+        # Update grant state based on the heartbeat response.
+        self._mapResponseCodeToGrantState(heartbeat_response['response']['responseCode'],
+                                         grant_object)
+
+      return heartbeat_requests, heartbeat_responses
+
+  def _mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+    """Maps the heartbeat response and update the grant state."""
+    if heartbeat_response_code == Grant.ResponseCodes.SUCCESS.value:
+      grant_object.authorized_in_last_heartbeat = True
+    elif heartbeat_response_code == Grant.ResponseCodes.TERMINATED_GRANT.value:
+      grant_object.is_terminated = True
+    elif heartbeat_response_code == Grant.ResponseCodes.SUSPENDED_GRANT.value:
+      grant_object.authorized_in_last_heartbeat = False
+    else:
+      logging.error('Unknown code=%s received in Heartbeat Response',heartbeat_response_code)
+      raise Exception('Unknown code received in Heartbeat Response')
+
+  def _constructGrantRequest(self,heartbeat_response):
+    """Construct unwrapped version of grant request."""
+    grant_request = {
+      'cbsdId': heartbeat_response['cbsdId'],
+      'operationParam': {
+        'maxEirp': heartbeat_response['operationParam']['maxEirp'],
+        'operationFrequencyRange': {
+          'lowFrequency': heartbeat_response['operationParam']['operationFrequencyRange']['lowFrequency'],
+          'highFrequency': heartbeat_response['operationParam']['operationFrequencyRange']['highFrequency']
+        }
+      }
+    }
+    return grant_request
+
+  def performHeartbeatAndUpdateGrants(self):
+    """Does heartbeat request for all active grants and updates the grants.
+
+    The following steps are performed by this function
+      1. Send heartbeat request on all grant of CBSD at once
+      2. If the heartbeat response contains operationParam then those grants
+         are relinquished.
+      3. New grant requests with the suggested operationParam values in the
+         heartbeat response is sent. Failures are ignored.
+      4. Heartbeat is performed again for all grants. Failures are ignored.
+    """
+    # Heartbeat on all active grants
+    heartbeat_requests = []
+    heartbeat_requests, heartbeat_responses = self.heartbeatForAllActiveGrants()
+    relinquishment_requests = []
+    grant_requests = []
+
+    # Construct relinquishment and grant for the heart beat response which have operation param.
+    for heartbeat_response, heartbeat_request in zip(heartbeat_responses, heartbeat_requests):
+        if heartbeat_response['response']['responseCode'] == Grant.ResponseCodes.TERMINATED_GRANT.value:
+           if 'operationParam' in heartbeat_response:
+             # Construct grant request for Terminated grants
+             grant_requests.append(self._constructGrantRequest(heartbeat_response))
+           else:
+             logging.error('Invalid Response since operation param not present in case of Terminated Grant')
```

Comment:
Providing the grant ID and CBSD ID from the heartbeat request in this error message will help in understanding what happened if this case occurs.

Reply:
Changed as per suggestion